src/api/baggage.rs

```diff
@@ -56,12 +73,44 @@ impl Baggage {
         K: Into<Key>,
         V: Into<Value>,
     {
-        self.inner.insert(key.into(), value.into())
+        self.inner
+            .insert(key.into(), (value.into(), Metadata::default()))
+            .map(|pair| pair.0)
+    }
+
+    /// Inserts a name-value pair into the baggage.
+    ///
+    /// Same with `insert`, if the name was not present, [`None`] will be returned.
+    /// If the name is present, the old value and metadata will be returned.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use opentelemetry::api::{Baggage, Value};
+    ///
+    /// let mut cc = Baggage::new();
+    /// let _ = cc.insert_with_metadata("my-name", "my-value", "test");
+    ///
+    /// assert_eq!(cc.get_with_metadata("my-name"), Some(&(Value::String("my-value".to_string()), "test".into())))
+    /// ```
+    pub fn insert_with_metadata<K, V, S>(
+        &mut self,
+        key: K,
+        value: V,
+        metadata: S,
+    ) -> Option<(Value, Metadata)>
+    where
+        K: Into<Key>,
+        V: Into<Value>,
+        S: Into<Metadata>,
+    {
+        self.inner
```

Comment:
`insert_with_metadata` and `insert` should both probably check https://w3c.github.io/baggage/#limits as well

Reply:
Good catch. Will add those checks